test/handshake-memfail.c

```diff
@@ -0,0 +1,219 @@
+                                                SSL_ERROR_NONE)))
+        goto end;
+
+    testresult = 1;
+
+end:
+    SSL_free(serverssl);
+    SSL_free(clientssl);
+    SSL_CTX_free(sctx);
+    SSL_CTX_free(cctx);
+
+    return testresult;
+}
+
+/**
+ * @brief run our workload to count the number of allocations we make.
+ *
+ * Creates a new OpenSSL library context and performs a test SSL/TLS
+ * handshake. The number of malloc operations is recorded and printed for
+ * diagnostic purposes.
+ *
+ * @return 1 if the handshake succeeds, 0 otherwise.
+ */
+static int test_record_alloc_counts(void)
+{
+    int ret;
+    OSSL_LIB_CTX *libctx;
+
+    libctx = OSSL_LIB_CTX_new();
+    if (!TEST_ptr(libctx))
+        return 0;
+
+    ret = do_handshake(libctx);
+
+    OSSL_LIB_CTX_free(libctx);
+    libctx = NULL;
+
+    return ret;
+}
+
+/**
+ * @brief run our workload to count the number of allocations we make.
+ *
+ * Creates a new OpenSSL library context and performs a test SSL/TLS
+ * handshake.
+ *
+ * Note this is exactly the same as test_record_alloc_counts with 1 difference
+ * The test always returns 1.  We do this because with allocation failures
+ * in effect, we can't expect things to work, so we always return success
+ * so that the test keeps running.
+ */
+static int test_alloc_failures(void)
+{
+    OSSL_LIB_CTX *libctx;
+
+    libctx = OSSL_LIB_CTX_new();
+    if (!TEST_ptr(libctx))
+        return 1;
+
+    do_handshake(libctx);
+
+    OSSL_LIB_CTX_free(libctx);
+    libctx = NULL;
+
+    return 1;
+}
+
+static int test_report_alloc_counts(void)
+{
+    CRYPTO_get_alloc_counts(&mcount, &rcount, &fcount);
+    /*
+     * Report our memory allocations from the count run
+     * NOTE: We report a number of allocations to skip here
+     * (the scount value).  These are the allocations that took
+     * place while the test harness itself was getting setup
+     * (i.e. calling OPENSSL_init_crypto/etc).  We can't fail
+     * those allocations as they will cause the test to fail before
+     * we have even run the workload.  So report them so we can
+     * allow them to function before we start doing any real testing
+     */
+    TEST_info("skip: %d count %d\n", scount, mcount - scount);
+    return 1;
+}
+
+int setup_tests(void)
+{
+    static char *opmode = NULL;
+    static char *certsdir = NULL;
+
+    if (!TEST_ptr(opmode = test_get_argument(0)))
+        goto err;
+
+    if (!TEST_ptr(certsdir = test_get_argument(1)))
+        goto err;
+
+    cert = test_mk_file_path(certsdir, "servercert.pem");
+    if (cert == NULL)
+        goto err;
+
+    privkey = test_mk_file_path(certsdir, "serverkey.pem");
+    if (privkey == NULL)
+        goto err;
+
+    if (!strcmp(opmode, "count")) {
+        CRYPTO_get_alloc_counts(&scount, &rcount, &fcount);
+        ADD_TEST(test_record_alloc_counts);
+        ADD_TEST(test_report_alloc_counts);
+    } else {
+        ADD_TEST(test_alloc_failures);
+    }
+    return 1;
```

Comment:
Also need to cleanup `cert` and `privkey` here

Reply:
why? they're created globally in setup_tests, and scrubbed in cleanup_tests on exit